Competition2017/src/competition/subsystems/agitator/AgitatorSubsystem.java

```diff
@@ -1,43 +1,135 @@
+        this.pidProperties = new PIDPropertyManager("AgitatorSpeed", propManager, assertionManager, 0, 0, 0, 0);
+
+        agitatorSpeedThresh = propManager.createPersistentProperty(side + "Agitator nominal speed thresh (TPC)", motor);
+        agitatorCurrentSpeed = propManager.createEphemeralProperty(side + "Agitator current speed (TPC)", 0);
+        agitatorTargetSpeed = propManager.createEphemeralProperty(side + "Agitator goal speed (TPC)", 0);
+        agitatorOutputPower = propManager.createEphemeralProperty(side + "Agitator voltage", 0);
+        intakePowerProperty = propManager.createPersistentProperty("Agitator intake power", 0.5);
+        ejectPowerProperty = propManager.createPersistentProperty("Agitator eject power", -0.5);
+        atSpeedProp = propManager.createEphemeralProperty("Is" + side + " Agitator at speed?", false);
+        agitatorTalonError = propManager.createEphemeralProperty(side + " Agitator speed error", 0);
+        enableAgitatorLogging = propManager.createEphemeralProperty("Is " + side + " Agitator logging info?", false);
+
+        this.agitatorMotor = factory.getCANTalonSpeedController(motor);
+
+        setUpMotor(agitatorMotor);
+        agitatorMotor.createTelemetryProperties(side + "Agitator motor", propManager);
+        agitatorMotor.setBrakeEnableDuringNeutral(false);
         agitatorMotor.setProfile(0);
         agitatorMotor.setControlMode(TalonControlMode.PercentVbus);
-        motorPowerProperty = propManager.createPersistentProperty("Agitator motor power", 0.5);
     }
-    
-    public void stop(){
+
+    private void setUpMotor(XCANTalon agitatorMotor) {
+        log.info("Configuring " + side + " Agitator motor");
+        agitatorMotor.setProfile(0);
+        agitatorMotor.setFeedbackDevice(FeedbackDevice.QuadEncoder);
+
+        log.info("Configuring voltages");
+        agitatorMotor.configNominalOutputVoltage(0,  -0);
+        agitatorMotor.configPeakOutputVoltage(12, -12);
+
+        updateMotorPidValues(agitatorMotor);
+        agitatorMotor.setControlMode(TalonControlMode.Speed);
         agitatorMotor.set(0);
     }
+
+    private void updateMotorPidValues(XCANTalon motor) {
+        motor.setP(pidProperties.getP());
+        motor.setI(pidProperties.getI());
+        motor.setD(pidProperties.getD());
+        motor.setF(pidProperties.getF());
+    }
+
+    public void setAgitatorPower(double power) {
+        agitatorMotor.ensureTalonControlMode(TalonControlMode.PercentVbus);
+        agitatorMotor.set(power);
+    }
+
+    public void setAgitatorTargetSpeed(double speed) {
+        agitatorMotor.ensureTalonControlMode(TalonControlMode.Speed);
+        agitatorTargetSpeed.set(speed);
+
+        updateMotorPidValues(agitatorMotor);
+        agitatorMotor.set(speed);
+    }
+
+    public double getSpeed() {
+        return agitatorMotor.getSpeed();
+    }
     
-    public void spinForwards(){
-        agitatorMotor.set(motorPowerProperty.get());
+    public boolean isAtSpeed() {
```

Comment:
The agitator will likely never be "at speed" while there are balls in the hopper.